pkg/scheduler/plugins/nodenumaresource/plugin.go

```diff
@@ -0,0 +1,438 @@
+	return nil
+}
+
+func (p *Plugin) PreFilterExtensions() framework.PreFilterExtensions {
+	return nil
+}
+
+func getPreFilterState(cycleState *framework.CycleState) (*preFilterState, *framework.Status) {
+	value, err := cycleState.Read(stateKey)
+	if err != nil {
+		return nil, framework.AsStatus(err)
+	}
+	state := value.(*preFilterState)
+	return state, nil
+}
+
+func (p *Plugin) Filter(ctx context.Context, cycleState *framework.CycleState, pod *corev1.Pod, nodeInfo *framework.NodeInfo) *framework.Status {
+	state, status := getPreFilterState(cycleState)
+	if !status.IsSuccess() {
+		return status
+	}
+	if state.skip {
+		return nil
+	}
+
+	node := nodeInfo.Node()
+	if node == nil {
+		return framework.NewStatus(framework.Error, "node not found")
+	}
+
+	// It's necessary to force node to have NodeResourceTopology and CPUTopology
+	// We must satisfy the user's CPUSet request. Even if some nodes in the cluster have resources,
+	// they cannot be allocated without valid CPU topology.
+	numaInfo := p.nodeInfoCache.getNodeNUMAInfo(node.Name)
+	if numaInfo == nil {
+		return framework.NewStatus(framework.UnschedulableAndUnresolvable, ErrMissingNodeResourceTopology)
+	}
+
+	numaInfo.lock.RLock()
+	defer numaInfo.lock.RUnlock()
+	if !numaInfo.cpuTopology.IsValid() {
+		return framework.NewStatus(framework.UnschedulableAndUnresolvable, ErrInvalidCPUTopology)
+	}
+
+	return nil
+}
+
+func (p *Plugin) Score(ctx context.Context, cycleState *framework.CycleState, pod *corev1.Pod, nodeName string) (int64, *framework.Status) {
+	state, status := getPreFilterState(cycleState)
+	if !status.IsSuccess() {
+		return 0, status
+	}
+	if state.skip {
+		return 0, nil
+	}
+
+	nodeInfo, err := p.handle.SnapshotSharedLister().NodeInfos().Get(nodeName)
+	if err != nil {
+		return 0, framework.NewStatus(framework.Error, fmt.Sprintf("getting node %q from Snapshot: %v", nodeName, err))
+	}
+	node := nodeInfo.Node()
+	if node == nil {
+		return 0, framework.NewStatus(framework.Error, "node not found")
+	}
+
+	// There is no need to force nodes to have a NodeResourceTopology during the scoring phase.
+	numaInfo := p.nodeInfoCache.getNodeNUMAInfo(nodeName)
+	if numaInfo == nil {
+		return 0, nil
+	}
+
+	numaInfo.lock.RLock()
+	defer numaInfo.lock.RUnlock()
+	if !numaInfo.cpuTopology.IsValid() {
+		return 0, nil
+	}
+
+	score := p.calcScore(state.numCPUsNeeded, state.resourceSpec, numaInfo)
+	return score, nil
+}
+
+func (p *Plugin) calcScore(numCPUsNeeded int, resourceSpec *extension.ResourceSpec, numaInfo *nodeNUMAInfo) int64 {
+	availableCPUs, allocated := getAvailableCPUsFunc(numaInfo)
+	acc := newCPUAccumulator(
+		numaInfo.cpuTopology,
+		availableCPUs,
+		allocated,
+		numCPUsNeeded,
+		false,
+		p.pluginArgs.NUMAAllocateStrategy,
+	)
+
+	var freeCPUs [][]int
+	if resourceSpec.PreferredCPUBindPolicy == schedulingconfig.CPUBindPolicyFullPCPUs {
+		if numCPUsNeeded <= numaInfo.cpuTopology.CPUsPerNode() {
+			freeCPUs = acc.freeCoresInNode(true)
```

Comment:
the naming `freeCoresInNode` is kind of confusing. it means the `CPUs in free cores of each node`

Reply:
Give suggestions when reviewing, such as how about freeCoresSplitByNode?

Reply:
IMO, we'd better add a concise prefix/suffix about the method returns. So `freeCPUsInCoreByNode` is a suggested option. Anyway, the naming is fine for an inner interface.